### PR TITLE
Extended Unity path detection to support Unity Hub

### PR DESCRIPTION
**Pull request by jared-improbable**
#### Description
Extended Unity path detection to support the Unity Hub, and make it more likely users won't have to do anything special to run.
Merged the End2End tests into the Unity project, since it depended directly on `UNITY_HOME`
New tools are based on .NET Core 🎊 

#### Tests
Manual runs of each scenario. TeamCity builds.

#### Documentation
In-code documentation.

#### Primary reviewers
@jamiebrynes7 @mattyoung-improbable

### CODE REVIEW

File: workers/unity/Assets/Playground/Editor/SnapshotGenerator/SnapshotGenerator.cs [hunk 1]
```diff
@@ -36,12 +36,6 @@ public struct Arguments
                 new WorkerAttributeSet(new Improbable.Collections.List<string> { UnityGameLogicType })
             });
 
-        private static readonly WorkerRequirementSet ClientSet = new WorkerRequirementSet(
```
Comment (jared-improbable) [inline workers/unity/Assets/Playground/Editor/SnapshotGenerator/SnapshotGenerator.cs]: Drive-by warning fix.

File: tools/FindUnity/Program.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+namespace FindUnity
+{
+    internal static class Program
+    {
+        // If SpecialFolder.ProgramFiles returns the (x86) variant instead of "Program Files",
+        // check that "Prefer 32 bit" is unticked in the project's Build settings.
+        private static readonly string ProgramFiles = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles);
+        private static readonly string UnityHubPath = Path.Combine(ProgramFiles, "Unity", "Hub");
+        private static readonly string UnityPath = Path.Combine(ProgramFiles, "Unity");
+        private static readonly string ImprobableUnityPath = Path.Combine(@"c:\", "Unity");
```
Comment (jessicafalk) [inline tools/FindUnity/Program.cs]: this will only work for windows. Could we already add the support for mac / paths for it as I'm currently checking everything around it

Comment (jared-improbable) [reply tools/FindUnity/Program.cs]: Yep, I'll do that tomorrow when I have my Mac again :)

File: tools/FindUnity/Program.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
+﻿using System;
```
Comment (jessicafalk) [inline tools/FindUnity/Program.cs]: it seems kind of an overhead to move this to a C#  project. Is the logic in the bash script really too much that we want to do that?

Comment (jared-improbable) [reply tools/FindUnity/Program.cs]: We ultimately have to stop depending on bash, since it's kind of a weird requirement to push on Windows developers.
I'd also rather we not write our build process and non-trivial tools (like this) twice in something terrible like Powershell/cmd for Windows, or something terrible like bash for Mac/Linux. So, my solution is to write it in C# once, and make the entry point to the program be a very simple Powershell/bash script, as appropriate to the platform.

File: workers/unity/Assets/Playground/Editor/SnapshotGenerator/SnapshotGenerator.cs (hunk 1 above)
Comment (jamiebrynes7) [reply workers/unity/Assets/Playground/Editor/SnapshotGenerator/SnapshotGenerator.cs]: Was this just because it wasn't being used?

Comment (jared-improbable) [reply workers/unity/Assets/Playground/Editor/SnapshotGenerator/SnapshotGenerator.cs]: Yep!

File: ci/includes/pinned-tools.sh [hunk 4]
```diff
@@ -1,33 +1,19 @@
...
-  else
-    echo "ERROR: Unity was not found in the default location. Please set the UNITY_HOME environment variable to where Unity ${UNITY_VERSION} is installed." >&2
-    exit 1
-  fi
-}
-
-UNITY_DIR="$(getUnityDir)"
-export UNITY_EXE="${UNITY_DIR}/Editor/Unity.exe"
+# Print the .NETCore version to aid debugging,
+# as well as ensuring that later calls to the tool don't print the welcome message on first run.
```
Comment (jessicafalk) [inline ci/includes/pinned-tools.sh]: :trollface:

File: ci/includes/pinned-tools.sh [hunk 5]
```diff
@@ -1,33 +1,19 @@
...
+dotnet --version
 
 export LINTER="cleanupcode.exe"
 
-export MSBUILD="$(powershell  –ExecutionPolicy Bypass ./ci/find-msbuild.ps1)"
+MSBUILD="$(powershell –ExecutionPolicy Bypass ./ci/find-msbuild.ps1)"
+export MSBUILD
```
Comment (jessicafalk) [inline ci/includes/pinned-tools.sh]: are we still using `MSBUILD`?

Comment (jared-improbable) [reply ci/includes/pinned-tools.sh]: Yep, until all of @jamiebrynes7 's PRs are merged

**Comment by jessicafalk** [comment]
@jamiebrynes7 don't worry, by the time you are back, Jared and I will have hopefully managed to merge everything in. ;)

**Comment by jared-improbable** [comment]
Added back in generated code (I used the exhaustive schema types)

File: schema/tests/component_data_test.schema [hunk 6]
```diff
@@ -1,13 +0,0 @@
-package tests;
```
Comment (jared-improbable) [inline schema/tests/component_data_test.schema]: This was completely unused.

File: schema/tests/component_test.schema [hunk 7]
```diff
@@ -1,13 +0,0 @@
-package tests;
```
Comment (jared-improbable) [inline schema/tests/component_test.schema]: This was completely unused.

File: ci/includes/pinned-tools.sh [hunk 8]
```diff
@@ -1,34 +1,18 @@
...
 
 DOTNET_VERSION="$(dotnet --version)"
 
 export MSBuildSDKsPath="${PROGRAMFILES}/dotnet/sdk/${DOTNET_VERSION}/Sdks"
+
+pushd workers/unity
+    UNITY_DIR="$(dotnet run -p ../../tools/FindUnity/FindUnity.csproj)"
+popd
+
+export UNITY_HOME="${UNITY_DIR}"
```
Comment (jessicafalk) [inline ci/includes/pinned-tools.sh]: let's remove this line. It shouldn't be needed anymore

Comment (jared-improbable) [reply ci/includes/pinned-tools.sh]: Whoops, bad merge!

File: ci/includes/pinned-tools.sh [hunk 9]
```diff
@@ -1,34 +1,15 @@
...
-    exit 1
-  fi
-}
-
-UNITY_DIR="$(getUnityDir)"
-export UNITY_HOME="${UNITY_DIR}"
-export UNITY_EXE="${UNITY_DIR}/Editor/Unity.exe"
+# Print the .NETCore version to aid debugging,
+# as well as ensuring that later calls to the tool don't print the welcome message on first run.
+dotnet --version
```
Comment (jessicafalk) [inline ci/includes/pinned-tools.sh]: do we still need this line, we are also running it in line 9 :)

File: tools/UnityPaths/UnityPaths.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+        public static string UnityHomeValue => Environment.GetEnvironmentVariable(UnityHomeEnvVar);
+
+        // Adjust the root based on Windows or MacOS. Linux is currently unsupported.
+        public static string ImprobableUnityRootPath =>
+            RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "/Applications" : Path.Combine(@"c:\", "Unity");
+
+        public static string TryGetUnityPath()
+        {
+            if (!Directory.Exists("Assets"))
```
Comment (jamiebrynes7) [inline tools/UnityPaths/UnityPaths.cs]: Is there a safer way to check we are in a Unity project? This seems kinda brittle?

Comment (jared-improbable) [reply tools/UnityPaths/UnityPaths.cs]: We could check for "ProjectSettings" as well. Otherwise, there's not anything else we can check for.
